src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -3011,34 +3351,47 @@ def _sync_runs(self) -> None:
         exists in the database, and if not, creates it.
         """
-        # Get all runs from ZenML.
+        # Sync all MLMD runs that don't exist in ZenML.
+        # For performance reasons, we determine this by checking the MLMD ID
+        # of the run, which we can do since MLMD returns runs in ascending
+        # order of their ID.
         with Session(self.engine) as session:
-            zenml_runs_list = session.exec(select(PipelineRunSchema)).all()
-        zenml_runs = {run.name: run.to_model() for run in zenml_runs_list}
+            synced_mlmd_ids = session.exec(
+                select(PipelineRunSchema.mlmd_id).where(
+                    isnot(PipelineRunSchema.mlmd_id, None)
+                )
+            ).all()
+        max_synced_mlmd_id = max(
+            [id_ for id_ in synced_mlmd_ids if id_ is not None], default=0
+        )
 
-        # Get all runs from MLMD.
-        mlmd_runs = self.metadata_store.get_all_runs()
+        unsynced_mlmd_runs = self.metadata_store.get_all_runs(
+            min_mlmd_id=max_synced_mlmd_id + 1
+        )
+        for mlmd_run in unsynced_mlmd_runs:
+            new_run = PipelineRunModel(
+                name=mlmd_run.name,
+                mlmd_id=mlmd_run.mlmd_id,
+                project=mlmd_run.project,
+                user=mlmd_run.user,
+                stack_id=mlmd_run.stack_id,
+                pipeline_id=mlmd_run.pipeline_id,
+                pipeline_configuration=mlmd_run.pipeline_configuration,
+                num_steps=mlmd_run.num_steps,
+                status=ExecutionStatus.RUNNING,
+            )
+            self.create_run(new_run)
 
-        # Sync all MLMD runs that don't exist in ZenML.
-        for run_name, mlmd_run in mlmd_runs.items():
-
-            # If the run is in MLMD but not in ZenML, we create it
-            if run_name not in zenml_runs:
-                new_run = PipelineRunModel(
-                    name=run_name,
-                    mlmd_id=mlmd_run.mlmd_id,
-                    project=mlmd_run.project,
-                    user=mlmd_run.user,
-                    stack_id=mlmd_run.stack_id,
-                    pipeline_id=mlmd_run.pipeline_id,
-                    pipeline_configuration=mlmd_run.pipeline_configuration,
-                    num_steps=mlmd_run.num_steps,
-                )
-                new_run = self._create_run(new_run)
-                zenml_runs[run_name] = new_run
-
-        for run_ in zenml_runs.values():
-            self._sync_run_steps(run_.id)
+        # Sync steps of all unfinished runs.
+        with Session(self.engine) as session:
+            unfinished_run_ids = session.exec(
+                select(PipelineRunSchema.id).where(
+                    PipelineRunSchema.status == ExecutionStatus.RUNNING
+                )
+            ).all()
+        for run_id in unfinished_run_ids:
+            self._sync_run_steps(run_id)
+            self.get_run(run_id)  # Updates run status if it has changed.
```

Comment:
This seems like an expensive call just to sync the status to the run. Can we maybe include that logic in the `sync_run_steps` if that is already doing all the queries?

Reply:
I like the idea of merging step and status sync into a single method `sync_run()`. It feels very clean at first glance. Unfortunately, after implementing it, I noticed that this leads to a difficult circular dependency:
- `sync_run()` calls `get_run()` to sync the status
- `get_run()` needs to have all steps to determine the status and accordingly needs to call `sync_run()`